pennylane_lightning/core/src/simulators/lightning_gpu/measurements/MeasurementsGPU.hpp

```diff
@@ -0,0 +1,762 @@
+        -> PrecisionT {
+        std::vector<CFP_t> matrix_cu(gate_matrix.size());
+
+        for (std::size_t i = 0; i < gate_matrix.size(); i++) {
+            matrix_cu[i] =
+                cuUtil::complexToCu<std::complex<PrecisionT>>(gate_matrix[i]);
+        }
+
+        if (gate_matrix.empty()) {
+            std::string message = "Currently unsupported observable";
+            throw LightningException(message.c_str());
+        }
+
+        // Wire order reversed to match expected custatevec wire ordering for
+        // tensor observables.
+        auto &&local_wires =
+            (gate_matrix.empty())
+                ? wires
+                : std::vector<size_t>{wires.rbegin(), wires.rend()};
+
+        auto expect_val = this->getExpectationValueDeviceMatrix_(
+            matrix_cu.data(), local_wires);
+        return expect_val;
+    }
+
+    /**
+     * @brief Get expectation of a given host or device defined array.
+     *
+     * @param matrix Host or device defined row-major order gate matrix array.
+     * @param tgts Target qubits.
+     * @return auto Expectation value.
+     */
+    auto
+    getExpectationValueDeviceMatrix_(const CFP_t *matrix,
+                                     const std::vector<std::size_t> &tgts) {
+        void *extraWorkspace = nullptr;
+        size_t extraWorkspaceSizeInBytes = 0;
+
+        std::vector<int> tgtsInt(tgts.size());
+        std::transform(tgts.begin(), tgts.end(), tgtsInt.begin(),
+                       [&](std::size_t x) {
+                           return static_cast<int>(
+                               this->_statevector.getNumQubits() - 1 - x);
+                       });
+
+        size_t nIndexBits = this->_statevector.getNumQubits();
+        cudaDataType_t data_type;
+        cudaDataType_t expectationDataType =
+            CUDA_C_64F; // Requested by the custatevecComputeExpectation API
+        custatevecComputeType_t compute_type;
+
+        if constexpr (std::is_same_v<CFP_t, cuDoubleComplex> ||
+                      std::is_same_v<CFP_t, double2>) {
+            data_type = CUDA_C_64F;
+            compute_type = CUSTATEVEC_COMPUTE_64F;
+        } else {
+            data_type = CUDA_C_32F;
+            compute_type = CUSTATEVEC_COMPUTE_32F;
+        }
+
+        // check the size of external workspace
+        PL_CUSTATEVEC_IS_SUCCESS(custatevecComputeExpectationGetWorkspaceSize(
+            /* custatevecHandle_t */ this->_statevector.getCusvHandle(),
+            /* cudaDataType_t */ data_type,
+            /* const uint32_t */ nIndexBits,
+            /* const void* */ matrix,
+            /* cudaDataType_t */ data_type,
+            /* custatevecMatrixLayout_t */ CUSTATEVEC_MATRIX_LAYOUT_ROW,
+            /* const uint32_t */ tgtsInt.size(),
+            /* custatevecComputeType_t */ compute_type,
+            /* size_t* */ &extraWorkspaceSizeInBytes));
+
+        if (extraWorkspaceSizeInBytes > 0) {
+            PL_CUDA_IS_SUCCESS(
```

Comment:
This can be ignored. Let's exclude it from coverage

Reply:
Done!